### PR TITLE
Make validators accept lambdas without record argument

### LINKED ISSUE #45117: ComparisonValidator example code does not work
### Steps to reproduce

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", "7.0.3"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :people, force: true do |t|
    t.date :birth_date
  end
end

class Person < ActiveRecord::Base
  validates_comparison_of :birth_date, less_than_or_equal_to: -> { Date.today }
end

class BugTest < Minitest::Test
  def test_comparison_validator
    person = Person.create!(birth_date: Date.yesterday)

    assert_equal 1, Person.count
  end
end
```

### Expected behavior

```
1 runs, 1 assertions, 0 failures, 0 errors, 0 skips
```

### Actual behavior

```
Error:
BugTest#test_comparison_validator:
ActiveRecord::RecordInvalid: Validation failed: Birth date wrong number of arguments (given 1, expected 0)

1 runs, 0 assertions, 0 failures, 1 errors, 0 skips
```

I tried example code.
https://github.com/rails/rails/blob/v7.0.3/activemodel/lib/active_model/validations/comparison.rb#L74=
```ruby
      # For example:
      #
      #   class Person < ActiveRecord::Base
      #     validates_comparison_of :birth_date, less_than_or_equal_to: -> { Date.today }
      #     validates_comparison_of :preferred_name, other_than: :given_name, allow_nil: true
      #   end
```

PR: https://github.com/rails/rails/pull/40095
There seems to be no test for `-> { }`.

This works.
```ruby
validates_comparison_of :birth_date, less_than_or_equal_to: Proc.new { Date.today }
```

### System configuration
**Rails version**: 7.0.3

**Ruby version**: 3.1.2

### PR DESCRIPTION
**Pull request by fatkodima**
Fixes #45117.

### CODE REVIEW

**Review by jonathanhefner — COMMENTED** [review]
It looks like this pattern also occurs elsewhere:

* [`active_model/validations/length.rb`](https://github.com/rails/rails/blob/fe8d41eda16911d2ef0e7d08bb02d16650b2dcdf/activemodel/lib/active_model/validations/length.rb#L48-L53)
* [`active_model/validations/numericality.rb`](https://github.com/rails/rails/blob/fe8d41eda16911d2ef0e7d08bb02d16650b2dcdf/activemodel/lib/active_model/validations/numericality.rb#L117-L124)
* [`active_model/validations/format.rb`](https://github.com/rails/rails/blob/fe8d41eda16911d2ef0e7d08bb02d16650b2dcdf/activemodel/lib/active_model/validations/format.rb#L28)
* [`active_model/validations/clusivity.rb`](https://github.com/rails/rails/blob/fe8d41eda16911d2ef0e7d08bb02d16650b2dcdf/activemodel/lib/active_model/validations/clusivity.rb#L19-L25)

I feel like we should be consistent in what we support, so what do you think about factoring out a `resolve_value(record, value)` method?  It could live in a `ResolveValue` module in `active_model/validations/resolve_value.rb`, which could be included where required.

(Alternatively, we could add it as a utility method to the `Validator` or `EachValidator` class, but that feels a little off.)

Note that in some cases we check `value.respond_to?(:call)` instead of `Proc === value`, so we would (unfortunately) probably have to check for both to avoid breaking existing code.

**Comment by fatkodima** [comment]
Ooh, I was sure that this is already supported in other places. Thanks for the suggestion, I will do that.

**Comment by fatkodima** [comment]
Updated.

File: activemodel/lib/active_model/validations/resolve_value.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        else
+          if value.respond_to?(:call)
+            if value.method(:call).arity == 0
+              value.call
+            else
+              value.call(record)
+            end
+          else
+            value
+          end
```
[the remark this answers is not in the document]
Comment (fatkodima) [reply activemodel/lib/active_model/validations/resolve_value.rb:25]: 😅 
All tests are passing without that check.  

> Even if we do support #call objects, though, I think we can require that they have non-zero arity.

Yes, it makes a little sense for me to use the whole `:call`able objects *without* arguments, when it is easier to just write a proc or lambda.

`:call`ables are not documented, but probably someone used them? So I suggest to support `:call`ables, but only with an argument (like before)?

[the remark this answers is not in the document]
Comment (jonathanhefner) [reply activemodel/lib/active_model/validations/resolve_value.rb:25]: Sounds reasonable to me. :+1:

**Comment by jonathanhefner** [comment]
Thank you, @fatkodima! :smile:

**Pull request merged**